tamr_toolbox/low_latency/llm.py

```diff
@@ -0,0 +1,215 @@
+    max_num_matches: Optional[int] = None,
+) -> Dict[int, List[JsonDict]]:
+    """
+    Find the best matching clusters or records for each supplied record. Returns a dictionary where
+    each key correpsonds to an input record and the value is a list of the LLM results for that
+    record. An empty result list indicates a null response from LLM (or no responses above the
+    min_match_prob, if that parameter was supplied).
+
+    Args:
+        match_client: a Tamr client set to use the port of the Match API
+        project_name: name of target mastering project
+        records: record or list of records to match
+        type: one of "records" or  "clusters" -- whether to pull record or cluster matches
+        primary_key: a primary key for the data; if supplied, this must be a field in input records
+        batch_size: split input into this batch size for LLM calls (e.g. to prevent network
+            timeouts), default None sends a single LLM call with all records
+        min_match_prob: if set, only matches with probability above minimum will be returned,
+            default None
+        max_num_matches: if set, at most max_num_matches will be returned for each input record in
+            records, default None
+
+    Returns:
+        Dict keyed by integers (indices of inputs), or by primary_key if primary_key is supplied,
+            with value a list containing matched data
+
+    Raises:
+        ValueError: if match type is not "records" or "clusters", or if batch_size is non-positive
+        RuntimeError: if query fails
+    """
+
+    result_dict = defaultdict(lambda: [])  # dict which defaults to empty list to hold results
+
+    url = f"/api/v1/projects/{project_name}:match?type={type}"
+
+    # Set up keys to read results
+    if type == "records":
+        record_key = "queryRecordId"
+        prob_key = "matchProbability"
+    elif type == "clusters":
+        record_key = "entityId"
+        prob_key = "avgMatchProb"
+    else:
+        raise ValueError(f"Unsupported match type {type}.")
+
+    # Convert single record to list for processing
+    if isinstance(records, Dict):
+        records = [records]
+
+    # Check batch size and set if not supplied
+    if batch_size is None:
+        batch_size = len(records)
+        if batch_size == 0:
+            LOGGER.warn("No input supplied to llm_query -- returning empty result.")
+            return result_dict
+    elif batch_size <= 0:
+        raise ValueError(f"Batch size must be non-negative: received {batch_size}")
+
+    # Split into batches and convert to LLM query format
+    for j in range(0, len(records), batch_size):
+        json_recs = _prepare_json(records[j : j + batch_size], primary_key=primary_key, offset=j)
+        response = match_client.post(url, json=json_recs)
+
+        # Process responses
+        if response.ok:
+            if response.content == b"":  # handle null response
+                continue
```

Comment:
No, the empty `b''` response comes with a 200 code, so `.successful` is true

Reply:
Is the `b''` response an error?

Reply:
or do you mean use that instead of just the `response.ok` line?

Reply:
Yeah, mostly to replace the `response.ok`.

Reply:
It's not an error, it just means the request went through but found no matches for any of the inputs

Reply:
Great!  Yeah I would just use the `.successful()` method rather than checking the `.ok` property to see if the request was successful.  Handling an empty return is still going to need to be custom.